FEXCore/Source/Interface/Core/Core.cpp

```diff
@@ -497,18 +498,33 @@ void ContextImpl::LockBeforeFork(FEXCore::Core::InternalThreadState* Thread) {
 }
 #endif
 
-void ContextImpl::ClearCodeCache(FEXCore::Core::InternalThreadState* Thread) {
+void ContextImpl::OnCodeBufferAllocated(CPU::CodeBuffer& Buffer) {
+  if (Config.GlobalJITNaming()) {
+    Symbols.RegisterJITSpace(Buffer.Ptr, Buffer.Size);
+  }
+}
+
+void ContextImpl::ClearCodeCache(FEXCore::Core::InternalThreadState* Thread, bool NewCodeBuffer) {
   FEXCORE_PROFILE_INSTANT("ClearCodeCache");
 
   if (CodeObjectCacheService) {
     // Ensure the Code Object Serialization service has fully serialized this thread's data before clearing the cache
     // Use the thread's object cache ref counter for this
     CodeSerialize::CodeObjectSerializeService::WaitForEmptyJobQueue(&Thread->ObjectCacheRefCounter);
   }
-  std::lock_guard<std::recursive_mutex> lk(Thread->LookupCache->WriteLock);
 
-  Thread->LookupCache->ClearCache();
-  Thread->CPUBackend->ClearCache();
+  if (NewCodeBuffer) {
+    // NOTE: Holding on to the reference here is required to ensure validity of the WriteLock mutex
+    std::shared_ptr PrevCodeBuffer = Thread->CPUBackend->GetThreadLocalCodeBuffer();
+    std::lock_guard lk(PrevCodeBuffer->LookupCache->WriteLock);
+
+    // Allocate new CodeBuffer + L3 LookupCache, then clear L1+L2 caches
+    Thread->CPUBackend->ClearCache();
```

Comment:
Could CPUBackend handle the mapping change itself? It already does on code compilation and it would allow CPUBackend->ClearCache to be called without worrying about updating the mapping exteernally